only update the container health status each time at the very beginning of deploy agent starts

liyaqin1 opened the pull request:
We need to send container health status to agent server so that Teletraan UI page can expose the container health status.

liyaqin1 commented:
Tested locally on mcrouter container only. I can see the extraInfo includes the json:
`PingRequest(hostId=i-0dd4b6ce5062c36f0, hostName=dev-yaqinli, hostIp=10.9.215.252, agentVersion=1.2.46-pinterest-1.0.2, autoscalingGroup=None, availabilityZone=None, stageType=None, groups=['singer-prod-sidecar', 'envoy-devapp', 'tcollector-prod-sidecar', 'mcrouter-devapp', 'mcrouter-prod-sidecar', 'metrics-agent-prod-sidecar', 'zk_update_monitor-prod-sidecar'], accountId=None, reports=PingReport(deployId=YINT0WznQEm_kLaUr9BGGQ, envId=nS2GIXkgQnK4pR9qG5ZpFg, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=8IBIIsjNQ9-lr7q1JrFnVw, envId=inbDfs6KQICpRRfuKQw_bQ, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=3zZuGjXbReqgsgwkAFcQPQ, envId=2XwshUOHR36DGEAH4PzBwg, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo={'Status': 'healthy', 'FailingStreak': 0, 'Log': [{'Start': '2023-09-16T02:04:44.689323523Z', 'End': '2023-09-16T02:04:44.807857464Z', 'ExitCode': 0, 'Output': ''}, {'Start': '2023-09-16T02:05:44.815379602Z', 'End': '2023-09-16T02:05:44.895910023Z', 'ExitCode': 0, 'Output': ''}, {'Start': '2023-09-16T02:06:44.905974845Z', 'End': '2023-09-16T02:06:44.993351896Z', 'ExitCode': 0, 'Output': ''}, {'Start': '2023-09-16T02:07:45.004011134Z', 'End': '2023-09-16T02:07:45.083730003Z', 'ExitCode': 0, 'Output': ''}, {'Start': '2023-09-16T02:08:45.091941857Z', 'End': '2023-09-16T02:08:45.167925972Z', 'ExitCode': 0, 'Output': ''}]}, deployAlias=None),PingReport(deployId=PfyxggJtQ9uD7yCgXUMtwA, envId=kwUx6aHWQo6R174lT-eawQ, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=CDBJf7s1Sh2G7V1mZVPhpA, envId=typrFRIFR6-lDyaiNjAk9w, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=rXqFmaubTIGROjX0XT4VAg, envId=gu2DEmL4Tv2XOqDKFc7Axw, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=V-2pBLZ9QMKG5gfoLR7xEg, envId=7ib3xSHSR3unnLdN-ViGxw, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=QH71hZjfSn2MheO8nmMIkA, envId=pZwvW9faQTW7vD19ER72Gg, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=stx01o7qQiKnKOFX2fL5Kg, envId=99xVHd7bSr608HSHqrCGbA, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=qXyg0uh1RpeHfUiTHo_Q8g, envId=tLWOA1CsQ4W2Nu_kMFyYVA, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None))`

ntascii commented on deploy-agent/deployd/agent.py:
```diff
@@ -120,6 +122,22 @@ def serve_build(self):
+            for status in self._envs.values():
+                # for each container, we check the health status
+                try:
+                    cmd = ['docker', 'inspect', status.report.envName]
```
I think we have an util to execute external cmd.

ntascii replied:
probably move it to a helper function in common/utils.py

liyaqin1 replied:
sounds good to me, thanks

ntascii commented on deploy-agent/deployd/agent.py:
```diff
@@ -120,6 +122,22 @@ def serve_build(self):
+                # for each container, we check the health status
+                try:
+                    cmd = ['docker', 'inspect', status.report.envName]
+                    output = subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout
```
you can use this cmd `docker inspect -f '{{.State.Health.Status}}'`

liyaqin1 replied:
sure, thanks

ntascii commented on deploy-agent/deployd/agent.py:
```diff
@@ -120,6 +122,22 @@ def serve_build(self):
         # start to ping server to get the latest deploy goal
+        if len(self._envs) > 0:
+            for status in self._envs.values():
+                # for each container, we check the health status
```
we only check health when there is no deployments to perform

liyaqin1 replied:
discussed offline, we agree that we can check the health when the deploy agent starts.

ntascii commented on deploy-agent/deployd/client/client.py:
```diff
@@ -216,7 +216,7 @@ def send_reports(self, env_reports=None):
                                         stageType=self._stage_type,
                                         accountId=self._account_id)
-
+                log.info('ping request: %s' % ping_request)
```
I think we already log ping request, any reason we log here as well?

liyaqin1 replied:
oh, it is because when I tested locally, I got json error when pinging the agent server since I have not updated on agent server side yet. In this case, I have to log here to see what is the ping request because the code cannot execute to the place that has the log for `ping request -> ping response`

ntascii commented on deploy-agent/deployd/agent.py:
```diff
@@ -120,6 +122,22 @@ def serve_build(self):
+                    output = subprocess.run(cmd, check=True, stdout=subprocess.PIPE).stdout
+                    result = json.loads(output)
+                    if result[0].get("State"):
+                        status.report.extraInfo = result[0].get("State").get("Health")
```
might want to use extraInfo={'serviceHealth': <health status of the container>} per the design doc.

liyaqin1 replied:
make sense

ntascii commented on deploy-agent/deployd/agent.py:
```diff
@@ -120,6 +122,22 @@ def serve_build(self):
+                    log.info('sidecar name: {}'.format(status.report.envName))
+                    log.info('sidecar extraInfo: {}'.format(status.report.extraInfo))
+                except Exception:
+                    log.error('get exception while trying to check container health: {}'.format(traceback.format_exc()))
```
log.exception()

liyaqin1 replied:
make sense, thank you

liyaqin1 commented:
After remove json.dumps for extraInfo, I get 200 response from server.
```
2023-09-18 23:54:43,254 deployd.client.client:219 INFO ping request: PingRequest(hostId=i-0dd4b6ce5062c36f0, hostName=dev-yaqinli, hostIp=10.9.215.252, agentVersion=1.2.46-pinterest-1.0.2, autoscalingGroup=None, availabilityZone=None, stageType=None, groups=['singer-prod-sidecar', 'envoy-devapp', 'tcollector-prod-sidecar', 'mcrouter-devapp', 'mcrouter-prod-sidecar', 'metrics-agent-prod-sidecar', 'zk_update_monitor-prod-sidecar'], accountId=None, reports=PingReport(deployId=tWH3VrCZRfakS2N5up1k2g, envId=inbDfs6KQICpRRfuKQw_bQ, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=3zZuGjXbReqgsgwkAFcQPQ, envId=2XwshUOHR36DGEAH4PzBwg, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo={'serviceHealth': 'healthy'}, deployAlias=None),PingReport(deployId=CDBJf7s1Sh2G7V1mZVPhpA, envId=typrFRIFR6-lDyaiNjAk9w, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=rXqFmaubTIGROjX0XT4VAg, envId=gu2DEmL4Tv2XOqDKFc7Axw, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None))
2023-09-18 23:54:43,255 urllib3.connectionpool:1018 DEBUG Starting new HTTPS connection (1): teletraan-agent.pinadmin.com:443
2023-09-18 23:54:43,362 urllib3.connectionpool:482 DEBUG https://teletraan-agent.pinadmin.com:443 "POST /v1/system/ping HTTP/1.1" 200 35
2023-09-18 23:54:43,362 deployd.client.client:224 DEBUG PingRequest(hostId=i-0dd4b6ce5062c36f0, hostName=dev-yaqinli, hostIp=10.9.215.252, agentVersion=1.2.46-pinterest-1.0.2, autoscalingGroup=None, availabilityZone=None, stageType=None, groups=['singer-prod-sidecar', 'envoy-devapp', 'tcollector-prod-sidecar', 'mcrouter-devapp', 'mcrouter-prod-sidecar', 'metrics-agent-prod-sidecar', 'zk_update_monitor-prod-sidecar'], accountId=None, reports=PingReport(deployId=tWH3VrCZRfakS2N5up1k2g, envId=inbDfs6KQICpRRfuKQw_bQ, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=3zZuGjXbReqgsgwkAFcQPQ, envId=2XwshUOHR36DGEAH4PzBwg, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo={'serviceHealth': 'healthy'}, deployAlias=None),PingReport(deployId=CDBJf7s1Sh2G7V1mZVPhpA, envId=typrFRIFR6-lDyaiNjAk9w, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None),PingReport(deployId=rXqFmaubTIGROjX0XT4VAg, envId=gu2DEmL4Tv2XOqDKFc7Axw, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo=None, deployAlias=None)) -> PingResponse(opCode=NOOP, deployGoal=None)
2023-09-18 23:54:43,363 deployd:373 INFO No deploy goal to be updated.
2023-09-18 23:54:43,363 deployd:195 INFO Complete the current deploy with response: PingResponse(opCode=NOOP, deployGoal=None).
```

liyaqin1 commented:
addressed the comments and tested again. looks good to me.
```
2023-09-19 00:53:03,228 urllib3.connectionpool:1018 DEBUG Starting new HTTPS connection (1): teletraan-agent.pinadmin.com:443
2023-09-19 00:53:03,298 urllib3.connectionpool:482 DEBUG https://teletraan-agent.pinadmin.com:443 "POST /v1/system/ping HTTP/1.1" 200 35
2023-09-19 00:53:03,299 deployd.client.client:224 DEBUG PingRequest(hostId=i-0dd4b6ce5062c36f0, hostName=dev-yaqinli, hostIp=10.9.215.252, agentVersion=1.2.46-pinterest-1.0.2, autoscalingGroup=None, availabilityZone=None, stageType=None, groups=['singer-prod-sidecar', 'envoy-devapp', 'tcollector-prod-sidecar', 'mcrouter-devapp', 'mcrouter-prod-sidecar', 'metrics-agent-prod-sidecar', 'zk_update_monitor-prod-sidecar'], accountId=None, reports=PingReport(deployId=tWH3VrCZRfakS2N5up1k2g, envId=inbDfs6KQICpRRfuKQw_bQ, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo={'serviceHealth': None}, deployAlias=None),PingReport(deployId=3zZuGjXbReqgsgwkAFcQPQ, envId=2XwshUOHR36DGEAH4PzBwg, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo={'serviceHealth': 'healthy'}, deployAlias=None),PingReport(deployId=CDBJf7s1Sh2G7V1mZVPhpA, envId=typrFRIFR6-lDyaiNjAk9w, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo={'serviceHealth': None}, deployAlias=None),PingReport(deployId=rXqFmaubTIGROjX0XT4VAg, envId=gu2DEmL4Tv2XOqDKFc7Axw, deployStage=SERVING_BUILD, status=SUCCEEDED, errorCode=0, errorMessage=None, failCount=0, extraInfo={'serviceHealth': None}, deployAlias=None)) -> PingResponse(opCode=NOOP, deployGoal=None)
2023-09-19 00:53:03,299 deployd:366 INFO No deploy goal to be updated.
2023-09-19 00:53:03,300 deployd:188 INFO Complete the current deploy with response: PingResponse(opCode=NOOP, deployGoal=None).
```

ntascii commented on deploy-agent/deployd/types/ping_request.py:
```diff
@@ -71,9 +71,11 @@ def to_json(self):
             ping_report["errorMessage"] = report.errorMessage
             ping_report["failCount"] = report.failCount
             ping_report["deployAlias"] = report.deployAlias
+            """
```
why this?

liyaqin1 replied:
If we use this, the server responds with "json error".

liyaqin1 replied:
since we didn't send any content for extraInfo before, I guess it is some code never tested. After I comment it out, the server responds 200 ok which means the map is accepted by java server.

ntascii commented on deploy-agent/tests/unit/deploy/server/test_agent.py:
```diff
@@ -60,37 +60,37 @@ def setUpClass(cls):
         cls.deploy_goal1 = {}
         cls.deploy_goal1['deployId'] = '123'
-        cls.deploy_goal1['envName'] = 'abc'
+        cls.deploy_goal1['envName'] = 'beacon'
```
can we revert the name of env back?

liyaqin1 replied:
if I don't change it to beacon which is a real container name, the test fails since I added some container check in the function

liyaqin1 replied:
the envName is the container name

ntascii replied:
have you considered mocking the container check in the test?
